egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py

```diff
@@ -411,4 +412,348 @@ def generate_pgru_config(self):
+                ans.append((config_name, line))
+        return ans
+
+    # convenience function to generate the PGRU config
+    def generate_pgru_config(self):
+
+        # assign some variables to reduce verbosity
+        name = self.name
+        # in the below code we will just call descriptor_strings as descriptors for conciseness
+        input_dim = self.descriptors['input']['dim']
+        input_descriptor = self.descriptors['input']['final-string']
+        cell_dim = self.config['cell-dim']
+        rec_proj_dim = self.config['recurrent-projection-dim']
+        nonrec_proj_dim = self.config['non-recurrent-projection-dim']
+        delay = self.config['delay']
+        repair_nonlin = self.config['self-repair-scale-nonlinearity']
+        repair_nonlin_str = "self-repair-scale={0:.10f}".format(repair_nonlin) if repair_nonlin is not None else ''
+        bptrunc_str = ("clipping-threshold={0}"
+                      " zeroing-threshold={1}"
+                      " zeroing-interval={2}"
+                      " recurrence-interval={3}"
+                      "".format(self.config['clipping-threshold'],
+                                self.config['zeroing-threshold'],
+                                self.config['zeroing-interval'],
+                                abs(delay)))
+        affine_str = self.config['ng-affine-options']
+        pes_str = self.config['ng-per-element-scale-options']
+
+        # Natural gradient per element scale parameters
+        # TODO: decide if we want to keep exposing these options
+        if re.search('param-mean', pes_str) is None and \
+           re.search('param-stddev', pes_str) is None:
+           pes_str += " param-mean=0.0 param-stddev=1.0 "
+
+        # string for GruNonlinearityComponent
+        gru_nonlin_str = ""
+        if self.config['param-stddev-nonlinearity'] != -1:
+            gru_nonlin_str += "param-stddev={0}".format(self.config['param-stddev-nonlinearity'])
+        if self.config['self-repair-threshold-nonlinearity'] != 0.2:
+            gru_nonlin_str += " self-repair-threshold={0}".format(self.config['self-repair-threshold-nonlinearity'])
+        if self.config['self-repair-scale-nonlinearity'] != 0.00001:
+            gru_nonlin_str += " self-repair-scale={0}".format(self.config['self-repair-scale-nonlinearity'])
+        
+        
+        # formulation like:
+        # z = \sigmoid ( x_t * U^z + s_{t-1} * W^z ) // update gate
+        # r = \sigmoid ( x_t * U^r + s_{t-1} * W^r ) // reset gate
+        # h = \tanh ( x_t * U^h + s_{t-1} \dot r * W^h )
+        # c_t = ( 1 - z ) \dot h + z \dot y_{t-1}
+        # y_t = c_t * W^s
+        # s_t = y_t(0:rec_proj_dim-1)
+        
+        
+
+        configs = []
+        configs.append("# W_zr matrics for z_t and r_t")
+        configs.append("component name={0}.W_zr type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + rec_proj_dim, cell_dim + rec_proj_dim, affine_str))
```

Comment:
You merge the two matrix into one, do you have to change the default 'max-change=0.75' ? increase it a little ?

Reply:
I will first change it to 1.0 and then do some experiments to determine the final value.

Reply:
@GaofengCheng Shall I need to change the default 'max-change=0.10', it seems that it doesn't work well for fast-pgru.